### DISCUSSION TITLE
Access bean from a pipe/adapter

### DISCUSSION THREAD

**Discussion by YummyToadies** [frankframework/frankframework#5646] [Q&A]
I want to fetch a setting through JMS from an external system and make this available as part of an xpathExpression on pipes (using property).

I tried now doing this through a bean, where I used hardcoded our own JMS QueueConnectionFactory and add it as a property using AppConstants. That works well, but I want to use the JMS connection detail from the resource from the Context.xml. I tried constructing JmsSender from the framework, but then it can't find the jmsRealm, because the resource is not registered with JNDI yet.

I was looking to init the bean lazy, but at some point it need to be initialized, but I don't know how to invoke this. That's why I was wondering if there is a way to do this. Or is a setup like this possible at all or is there an alternative way?

Especially since it needs to be available before the adapters are processed to replace the property with the real value.

**Comment by nielsm5** [comment]
A bit late, but I'm not quite understanding what you're trying to achieve here. You can look at our `test` module on how to setup the JNDI and connect to queues.

**Reply by YummyToadies** [reply]
We have a setting that I want to use in a pipe by using a property ( using syntax ${setting} ).
As far I know I need to fetch this before the configuration file is processed. Otherwise it won't be able to parse the ${setting} syntax.
Thus, my bean needs to run before this is done.
I was hoping to use the connectionFactory resource from the context.xml or using a JMS pipe.

Or, invoke the bean that fetches the setting from a pipe.

Example of how I want to use it
`            
<pipe name="checkValidRecord" className="'org.frankframework.pipes.XmlIf"
           getInputFromSessionKey="originalMessage"
           xpathExpression="root/setting= ${setting}">
        <forward name="then" path="next"/>
          <forward name="else" path="READY"/>
</pipe>
`


I 'fixed' it now by declaring bean in the springCustom.xml, pass the JMS connection string and setup jms connection in a custom java class.

Only problem I am running into now is how to mock the beans for Larva tests.

**Reply by nielsm5** [reply]
Mocking JMS for larva is very difficult, usually people stub the listeners/senders. Is there a reason you're not using mocks?
It's indeed possible to create mocks/jms connections using the springCustom.xml.

**Reply by YummyToadies** [reply]
I figured indeed it's not easy.
Stubbing the jms related pipes is good to use.

Not familiar with mocks using the beans in xml, I will figure that out as that is not something special for this framework.

As my current setup is working now, I will close the discussion.
Thx!
